internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go

```diff
@@ -25,205 +23,164 @@ import (
 	if err != nil {
 		return nil, err
 	} else if dec == nil {
-		f.ComputedRoutesCache.UntrackComputedRoutes(id)
+		f.cache.UntrackComputedRoutes(id)
 	}
 
 	return dec, err
 }
 
 func (f *Fetcher) FetchExplicitDestinationsData(
 	ctx context.Context,
-	explDestRefs []intermediateTypes.CombinedDestinationRef,
-) ([]*intermediateTypes.Destination, map[string]*intermediateTypes.Status, error) {
-	var (
-		destinations []*intermediateTypes.Destination
-		statuses     = make(map[string]*intermediateTypes.Status)
-	)
-
-	for _, dest := range explDestRefs {
-		// Fetch Destinations resource if there is one.
-		us, err := f.FetchDestinations(ctx, dest.ExplicitDestinationsID)
-		if err != nil {
-			// If there's an error, return and force another reconcile instead of computing
-			// partial proxy state.
-			return nil, statuses, err
-		}
+	proxyID *pbresource.ID,
+) ([]*intermediateTypes.Destination, *intermediateTypes.Status, error) {
 
-		if us == nil {
-			// If the Destinations resource is not found, then we should delete it from cache and continue.
-			f.DestinationsCache.DeleteDestination(dest.ServiceRef, dest.Port)
-			continue
-		}
+	var destinations []*intermediateTypes.Destination
 
+	// Fetch computed explicit destinations first.
+	cdID := resource.ReplaceType(pbmesh.ComputedDestinationsType, proxyID)
+	cd, err := resource.GetDecodedResource[*pbmesh.ComputedDestinations](ctx, f.client, cdID)
+	if err != nil {
+		return nil, nil, err
+	}
+	if cd == nil {
+		f.cache.UntrackComputedDestinations(cdID)
+		return nil, nil, nil
+	}
+
+	// Otherwise, track this resource in the destinations cache.
+	f.cache.TrackComputedDestinations(cd)
+
+	status := &intermediateTypes.Status{
+		ID:         cd.GetResource().GetId(),
+		OldStatus:  cd.GetResource().GetStatus(),
+		Generation: cd.GetResource().GetGeneration(),
+	}
+
+	for _, dest := range cd.GetData().GetDestinations() {
 		d := &intermediateTypes.Destination{}
 
 		var (
-			serviceID    = resource.IDFromReference(dest.ServiceRef)
-			serviceRef   = resource.ReferenceToString(dest.ServiceRef)
-			upstreamsRef = resource.IDToString(us.Resource.Id)
+			serviceID  = resource.IDFromReference(dest.DestinationRef)
+			serviceRef = resource.ReferenceToString(dest.DestinationRef)
 		)
 
 		// Fetch Service
 		svc, err := f.FetchService(ctx, serviceID)
 		if err != nil {
-			return nil, statuses, err
+			return nil, status, err
 		}
 
 		if svc == nil {
 			// If the Service resource is not found, then we update the status
-			// of the Upstreams resource but don't remove it from cache in case
-			// it comes back.
-			updateStatusCondition(statuses, upstreamsRef, dest.ExplicitDestinationsID,
-				us.Resource.Status, us.Resource.Generation, ctrlStatus.ConditionDestinationServiceNotFound(serviceRef))
+			// of the ComputedDestinations resource.
+			status.Conditions = append(status.Conditions, ctrlStatus.ConditionDestinationServiceNotFound(serviceRef))
```

Comment:
We still need to update status conditions on the computed destinations resources to:

1. wake up explicit destinations controller to re-reconcile it and update the status if needed. This is because that controller is not watching services or computed routes and so it won't update status on the Destinations resource if a service has changed for example.
2. We need to tell the user why we skipped this destination

Reply:
I guess alternatively the destinations controller could just look for the existence of the ComputedRoutes resource and use that as a signal, as you can't have one of those without a mesh-enabled, existing Service.

Reply:
Because even if that controller watches it, this controller will still need to validate it again because there's a possibility that things have changed between the reconcile of the computed destinations controller and this reconcile. So I thought since you can't really avoid this validation and you will end up with more watches overall on both controllers, this might be the best compromise.

Reply:
I removed all statuses for now and just made this controller skip the invalid destinations.